File: python/enolib/errors/validation.py
from ..error_types import ValidationError
from .selections import cursor, DOCUMENT_BEGIN, select_comments, select_element, select_key, selection
from ..constants import (
    ATTRIBUTE,
    BEGIN,
    EMBED_BEGIN,
    END,
    DOCUMENT,
    FIELD,
    ITEM
)


class Validation:
    @staticmethod
    def comment_error(context, message, element):
        return ValidationError(
            context.messages.comment_error(message),
            context.reporter(context).report_comments(element).snippet(),
            select_comments(element)
        )

    @staticmethod
    def element_error(context, message, element):
        return ValidationError(
            message,
            context.reporter(context).report_element(element).snippet(),
            select_element(element)
        )

    @staticmethod
    def key_error(context, message, element):
        return ValidationError(
            context.messages.key_error(message),
            context.reporter(context).report_line(element).snippet(),
            select_key(element)
        )

    @staticmethod
    def missing_comment(context, element):
        return ValidationError(
            context.messages.missing_comment,
            context.reporter(context).report_line(element).snippet(), # TODO: Question-tag an empty line before an element with missing comment
            selection(element, 'line', BEGIN)
        )

    @staticmethod
    def missing_element(context, key, parent, message):
        message = getattr(context.messages, message if key is None else message + '_with_key')

        return ValidationError(
            message if key is None else message(key),
            context.reporter(context).report_missing_element(parent).snippet(),
            DOCUMENT_BEGIN if parent['type'] == DOCUMENT else selection(parent, 'line', END)
        )

    @staticmethod
    def missing_value(context, element):
        selection_data = {}

        if (element['type'] == FIELD or
            element['type'] == EMBED_BEGIN):
            message = context.messages.missing_field_value(element['key'])

            if 'field_operator' in element['ranges']:
                selection_data['from'] = cursor(element, 'field_operator', END)
            else:
                selection_data['from'] = cursor(element, 'line', END)
        elif element['type'] == ATTRIBUTE:
            message = context.messages.missing_attribute_value(element['key'])
            selection_data['from'] = cursor(element, 'attribute_operator', END)
        elif element['type'] == ITEM:
            message = context.messages.missing_item_value(element['parent']['key'])
            selection_data['from'] = cursor(element, 'item_operator', END)

        snippet = context.reporter(context).report_element(element).snippet()

        if element['type'] == FIELD and 'continuations' in element:
            selection_data['to'] = cursor(element['continuations'][-1], 'line', END)
        else:
            selection_data['to'] = cursor(element, 'line', END)

        return ValidationError(message, snippet, selection_data)

    @staticmethod
    def unexpected_element(context, message, element):
        return ValidationError(
            message or context.messages.unexpected_element,
            context.reporter(context).report_element(element).snippet(),
            select_element(element)
        )
        
    @staticmethod
    def unexpected_field_content(context, key, field, message):
        message = getattr(context.messages, message if key is None else message + '_with_key')
        return ValidationError(
            message if key is None else message(key),
            context.reporter(context).report_element(field).snippet(),
            select_element(field)
        )

    @staticmethod
    def unexpected_multiple_elements(context, key, elements, message):
        message = getattr(context.messages, message if key is None else message + '_with_key')
        return ValidationError(
            message if key is None else message(key),
            context.reporter(context).report_elements(elements).snippet(),
            select_element(elements[0])
        )

    @staticmethod
    def unexpected_element_type(context, key, section, message):
        message = getattr(context.messages, message if key is None else message + '_with_key')

        return ValidationError(
            message if key is None else message(key),
            context.reporter(context).report_element(section).snippet(),
            select_element(section)
        )

    @staticmethod
    def value_error(context, message, element):
        if element['type'] == EMBED_BEGIN:
            if 'lines' in element:
                snippet = context.reporter(context).report_multiline_value(element).snippet()
                select = selection(element['lines'][0], 'line', BEGIN, element['lines'][-1], 'line', END)
            else:
                snippet = context.reporter(context).report_element(element).snippet()
                select = selection(element, 'line', END)
        else:
            snippet = context.reporter(context).report_element(element).snippet()
            select = {}

            if 'value' in element['ranges']:
                select['from'] = cursor(element, 'value', BEGIN)
            elif 'field_operator' in element['ranges']:
                select['from'] = cursor(element, 'field_operator', END)
            elif 'attribute_operator' in element['ranges']:
                select['from'] = cursor(element, 'attribute_operator', END)
            elif element['type'] == ITEM:
                select['from'] = cursor(element, 'item_operator', END)
            else:
                # TODO: Possibly never reached - think through state permutations
                select['from'] = cursor(element, 'line', END)

            if 'continuations' in element:
                select['to'] = cursor(element['continuations'][-1], 'line', END)
            elif 'value' in element['ranges']:
                select['to'] = cursor(element, 'value', END)
            else:
                select['to'] = cursor(element, 'line', END)

        return ValidationError(context.messages.value_error(message), snippet, select)

File: ruby/lib/enolib/errors/validation.rb
# frozen_string_literal: true

module Enolib
  module Errors
    module Validation
      def self.comment_error(context, message, element)
        ValidationError.new(
          context.messages.comment_error(message),
          context.reporter.new(context).report_comments(element).snippet,
          Selections.select_comments(element)
        )
      end

      def self.element_error(context, message, element)
        ValidationError.new(
          message,
          context.reporter.new(context).report_element(element).snippet,
          Selections.select_element(element)
        )
      end

      def self.key_error(context, message, element)
        ValidationError.new(
          context.messages.key_error(message),
          context.reporter.new(context).report_line(element).snippet,
          Selections.select_key(element)
        )
      end

      def self.missing_comment(context, element)
        ValidationError.new(
          context.messages::MISSING_COMMENT,
          context.reporter.new(context).report_line(element).snippet, # TODO: Question-tag an empty line before an element with missing comment
          Selections.selection(element, :line, RANGE_BEGIN)
        )
      end

      def self.missing_element(context, key, parent, message)
        ValidationError.new(
          key ? context.messages.send(message + '_with_key', key) : context.messages.const_get(message.upcase), # TODO: Solve the upcase rather through a different generated message type, e.g. method instead of constant
          context.reporter.new(context).report_missing_element(parent).snippet,
          parent[:type] == :document ? Selections::DOCUMENT_BEGIN : Selections.selection(parent, :line, RANGE_END)
        )
      end

      def self.missing_value(context, element)
        selection = {}

        if element[:type] == :field ||
           element[:type] == :embed_begin
          message = context.messages.missing_field_value(element[:key])

          selection[:from] =
            if element[:ranges].has_key?(:field_operator)
              Selections.cursor(element, :field_operator, RANGE_END)
            else
              Selections.cursor(element, :line, RANGE_END)
            end
        elsif element[:type] == :attribute
          message = context.messages.missing_attribute_value(element[:key])
          selection[:from] = Selections.cursor(element, :attribute_operator, RANGE_END)
        elsif element[:type] == :item
          message = context.messages.missing_item_value(element[:parent][:key])
          selection[:from] = Selections.cursor(element, :item_operator, RANGE_END)
        end

        snippet = context.reporter.new(context).report_element(element).snippet

        selection[:to] =
          if element[:type] == :field && element.has_key?(:continuations)
            Selections.cursor(element[:continuations].last, :line, RANGE_END)
          else
            Selections.cursor(element, :line, RANGE_END)
          end

        ValidationError.new(message, snippet, selection)
      end

      def self.unexpected_element(context, message, element)
        ValidationError.new(
          message || context.messages::UNEXPECTED_ELEMENT,
          context.reporter.new(context).report_element(element).snippet,
          Selections.select_element(element)
        )
      end
      
      def self.unexpected_field_content(context, key, field, message)
        ValidationError.new(
          key ? context.messages.send(message + '_with_key', key) : context.messages.const_get(message.upcase), # TODO: Solve the upcase rather through a different generated message type, e.g. method instead of constant
          context.reporter.new(context).report_element(field).snippet,
          Selections.select_element(field)
        )
      end

      def self.unexpected_multiple_elements(context, key, elements, message)
        ValidationError.new(
          key ? context.messages.send(message + '_with_key', key) : context.messages.const_get(message.upcase), # TODO: Solve the upcase rather through a different generated message type, e.g. method instead of constant
          context.reporter.new(context).report_elements(elements).snippet,
          Selections.select_element(elements[0])
        )
      end

      def self.unexpected_element_type(context, key, section, message)
        ValidationError.new(
          key ? context.messages.send(message + '_with_key', key) : context.messages.const_get(message.upcase), # TODO: Solve the upcase rather through a different generated message type, e.g. method instead of constant
          context.reporter.new(context).report_element(section).snippet,
          Selections.select_element(section)
        )
      end

      def self.value_error(context, message, element)
        if element[:type] == :embed_begin
          if element.has_key?(:lines)
            snippet = context.reporter.new(context).report_multiline_value(element).snippet
            select = Selections.selection(element[:lines][0], :line, RANGE_BEGIN, element[:lines][-1], :line, RANGE_END)
          else
            snippet = context.reporter.new(context).report_element(element).snippet
            select = Selections.selection(element, :line, RANGE_END)
          end
        else
          snippet = context.reporter.new(context).report_element(element).snippet
          select = {
            from:
              if element[:ranges].has_key?(:value)
                Selections.cursor(element, :value, RANGE_BEGIN)
              elsif element[:ranges].has_key?(:field_operator)
                Selections.cursor(element, :field_operator, RANGE_END)
              elsif element[:ranges].has_key?(:attribute_operator)
                Selections.cursor(element, :attribute_operator, RANGE_END)
              elsif element[:type] == :item
                Selections.cursor(element, :item_operator, RANGE_END)
              else
                # TODO: Possibly never reached - think through state permutations
                Selections.cursor(element, :line, RANGE_END)
              end,
            to:
              if element.has_key?(:continuations)
                Selections.cursor(element[:continuations][-1], :line, RANGE_END)
              elsif element[:ranges].has_key?(:value)
                Selections.cursor(element, :value, RANGE_END)
              else
                Selections.cursor(element, :line, RANGE_END)
              end
          }
        end

        ValidationError.new(context.messages.value_error(message), snippet, select)
      end
    end
  end
end
